File: Math/CPP/day_of_the_year.cpp
#include <string>
#include <vector>
using namespace std;

class Solution {
public:
    int dayOfYear(string date) {
        // Days in each month
        int days_freq[13] = { 31, 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31 };
        int year = stoi(date.substr(0, 4)),
            month = stoi(date.substr(5, 2)),
            day = stoi(date.substr(8, 2));
        if (isLeapYear(year)) days_freq[2]++;
        int sum = 0;
        for (int i = 0; i < month - 1; i++) sum += days_freq[i];
        return sum + min(sum, days_freq[month]);
    }

    bool isLeapYear(const int year) {
        return (year % 4 == 0 && year % 100 != 0) || (year % 400 == 0 && year % 4 == 0 && year % 100 == 0);
    }
};

File: Math/Python/day_of_the_year.py
class Solution:
    def dayOfYear(self, date: str) -> int:
        days_freq = [0, 31, 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31]
        y, m, d = map(int, date.split('-'))

        if self.is_leap_year(y):
            days_freq[2] = days_freq[2] + 1

        ndays = 0
        for i in range(m):
                ndays += days_freq[i]

        return ndays + min(days_freq[m], d)

    def is_leap_year(self, y: int) -> bool:
        return (y % 4 == 0 and y % 100 != 0) or (y % 4 == 0 and y % 100 == 0 and y % 400 == 0)